crypto/crypto/src/noise.rs

```diff
@@ -190,55 +251,69 @@ impl NoiseConfig {
         &self,
         rng: &mut (impl rand::RngCore + rand::CryptoRng),
         prologue: &[u8],
-        remote_public: &x25519::PublicKey,
+        remote_public: x25519::PublicKey,
         payload: Option<&[u8]>,
-    ) -> Result<(InitiatorHandshakeState, Vec<u8>), NoiseError> {
+        response_buffer: &mut [u8],
+    ) -> Result<InitiatorHandshakeState, NoiseError> {
         // checks
+        let mut expected_response_buffer_len = HANDSHAKE_MSG1_SIZE;
         if let Some(payload) = payload {
-            if payload.len()
-                > MAX_SIZE_NOISE_MSG - 2 * X25519_PUBLIC_KEY_LENGTH - 2 * AES_GCM_TAGLEN
-            {
+            expected_response_buffer_len += payload.len();
+            if payload.len() > MAX_SIZE_NOISE_MSG - HANDSHAKE_MSG1_SIZE {
                 return Err(NoiseError::PayloadTooLarge);
             }
         }
-
+        if response_buffer.len() < expected_response_buffer_len {
+            return Err(NoiseError::ResponseBufferTooSMall);
+        }
         // initialize
         let mut h = PROTOCOL_NAME.to_vec();
         let mut ck = PROTOCOL_NAME.to_vec();
+        let rs = remote_public; // for naming consistency with the specification
```

Comment:
(doing this so that one can more easily follow the spec when auditing this part of the code)